### PR TITLE
Add Monoid instance for RE

### LINKED ISSUE #37: Semigroup instance?
I've found with the -text variant of this package, when doing sophisticated replacements, it's neat to have a Semigroup instance and a helper `capture` function:
```haskell
capture = fmap snd . withMatched

instance Semigroup (RE' Text) where
  (<>) x y = (<>) <$> x <*> y
```

For example, I'm doing a script to rename modules, it looks like this:

```haskell
applyToContents config = replace inHeader
  where
    inHeader =
      capture (many (psym (\c -> elem c ['\n', '\r']))) <> string "module" <>
      capture (some (psym isSpace)) <>
      mname <>
      capture (some (psym isSpace))
    mname = string (configFrom config) *> pure (configTo config)
```

Without the `Semigroup`, I have to use `<*>` and `<$>` then `<>` to combine them.

Just throwing these out there. Using the returned value of the `Applicative` to do the replacement is totally awesome, by the way. Intuitive yet initially surprising.

### PR DESCRIPTION
**Pull request by chris-martin**
Resolves #37

### CODE REVIEW

File: Text/Regex/Applicative/Types.hs [hunk 1]
```diff
@@ -58,6 +58,11 @@ data Greediness = Greedy | NonGreedy
 -- and returns the list of @ra@'s return values on those strings.
 --
 -- * 'catMaybes' @ram@ matches iff @ram@ matches and produces 'Just _'.
+--
+-- * @ra@ '<>' @rb@ matches @ra@ followed by @rb@. The return value is @a <> b@,
+-- where @a@ and @b@ are the return values of @ra@ and @rb@ respectively.
+--
+-- * 'mempty' matches the empty string (i.e. it does not consume any symbols).
```
Comment (feuerbach) [inline Text/Regex/Applicative/Types.hs:65]: Perhaps also add that it returns `mempty` of the underlying type?

Comment (chris-martin) [reply Text/Regex/Applicative/Types.hs]: Okay, added.

File: Text/Regex/Applicative/Types.hs [hunk 2]
```diff
@@ -62,7 +62,8 @@ data Greediness = Greedy | NonGreedy
 -- * @ra@ '<>' @rb@ matches @ra@ followed by @rb@. The return value is @a <> b@,
 -- where @a@ and @b@ are the return values of @ra@ and @rb@ respectively.
 --
--- * 'mempty' matches the empty string (i.e. it does not consume any symbols).
+-- * 'mempty' matches the empty string (i.e. it does not consume any symbols),
+-- and its return value is the string type's 'mempty' value.
```
Comment (feuerbach) [inline Text/Regex/Applicative/Types.hs:66]: Why "the string type"? The string type is `[s]`, but we are talking about `a` here.

Comment (chris-martin) [reply Text/Regex/Applicative/Types.hs:66]: Ah, right. I was thinking of `RE s [s]` as the "typical" parameterization.

Comment (chris-martin) [reply Text/Regex/Applicative/Types.hs]: How would you prefer to resolve the Semigroup problem? Add a `semigroups` dependency, a conditional instance, or drop older GHC support?

Comment (feuerbach) [reply Text/Regex/Applicative/Types.hs]: If I'm reading https://hackage.haskell.org/package/base/changelog correctly, GHC >= 8 already have Data.Semigroup bundled. I think you may be simply missing the right imports.

**Comment by feuerbach** [comment]
Merged and released as v0.3.4, thanks.

**Pull request closed without merging**